datafusion/physical-plan/src/sorts/sort_preserving_merge.rs

```diff
@@ -98,19 +99,56 @@ pub struct SortPreservingMergeExec {
     ///
     /// See [`Self::with_round_robin_repartition`] for more information.
     enable_round_robin_repartition: bool,
+    ///
+    progressive_eval_exec: Option<Arc<ProgressiveEvalExec>>,
 }
 
 impl SortPreservingMergeExec {
     /// Create a new sort execution plan
     pub fn new(expr: LexOrdering, input: Arc<dyn ExecutionPlan>) -> Self {
         let cache = Self::compute_properties(&input, expr.clone());
+
+        // Todo: check if partition statistic is accurate
+        // Organize the input partitions into chains,
+        // where elements of each chain are input partitions that are
+        // non-overlapping, and each chain is ordered internally by their min/max statistics.
+        let partition_groups = input
+            .statistics_by_partition()
+            .and_then(|stats| {
+                let min_max =
+                    MinMaxStatistics::new_from_statistics(&expr, &input.schema(), &stats);
+                min_max
+            })
+            .map(|min_max_stats| {
+                let res = min_max_stats.first_fit();
```

Comment:
The binpacking into nonoverlapping chains occurs here.

We do something a bit different. We are ungrouping at the data source level such that each partition (and it's min/max range) can then we calculated up to the top. Since we poll based on each partition idx, we are not worried about re-grouping (since progressive eval knows the order in which to re-run through partitions).

Reply:
Note that I'm not sure about our current approach to upgrouping at data source, in order to maximize possible nonoverlapping ranges. It works for the plans we usually build, but may not be the best solution for the general approach.